src/TLAuctionHouse.sol

```diff
@@ -0,0 +1,563 @@
+    ///     - contract is not paused
+    ///     - block.timestamp is greater than the auction open timestamp
+    ///     - bid meets or exceeds the reserve price / min bid price
+    ///     - msg.sender has attached enough eth/erc20 as specified by `amount`
+    ///     - protocol fee has been supplied, if needed
+    /// @param nftAddress The nft contract address
+    /// @param tokenId The nft token id
+    /// @param amount The amount to bid in the currency address set in the auction
+    function bid(address nftAddress, uint256 tokenId, uint256 amount) external payable whenNotPaused nonReentrant {
+        // cache items
+        Auction memory auction = _auctions[nftAddress][tokenId];
+        IERC721 nft = IERC721(nftAddress);
+        bool firstBid;
+        bool durationExtended;
+
+        // check if the auction has started
+        if (auction.seller == address(0)) revert AuctionNotConfigured();
+        if (block.timestamp < auction.auctionOpenTime) revert AuctionNotOpen();
+
+        if (auction.startTime == 0) {
+            // first bid
+            // - check bid amount
+            // - clear sale
+            // - start the auction
+            // - escrow the NFT
+            if (amount < auction.reservePrice) revert BidTooLow();
+            delete _sales[nftAddress][tokenId];
+            auction.startTime = block.timestamp;
+            nft.transferFrom(auction.seller, address(this), tokenId);
+            firstBid = true;
+        } else {
+            // subsequent bids
+            // - check if auction ended
+            // - check bid amount
+            // - refund previous bidder
+            if (block.timestamp > auction.startTime + auction.duration) revert AuctionEnded();
+            if (amount < _calcNextMinBid(auction.highestBid)) revert BidTooLow();
+            uint256 refundAmount = auction.highestBid + _calcProtocolFee(auction.highestBid);
+            if (auction.currencyAddress == address(0)) {
+                _safeTransferETH(auction.highestBidder, refundAmount, weth);
+            } else {
+                _safeTransferERC20(auction.highestBidder, auction.currencyAddress, refundAmount);
+            }
+        }
+
+        // set highest bid
+        auction.highestBid = amount;
+        auction.highestBidder = msg.sender;
+
+        // extend auction if needed
+        uint256 timeRemaining = auction.startTime + auction.duration - block.timestamp;
+        if (timeRemaining < EXTENSION_TIME) {
+            auction.duration += EXTENSION_TIME - timeRemaining;
+            durationExtended = true;
+        }
+
+        // store updated parameters to storage
+        Auction storage sAuction = _auctions[nftAddress][tokenId];
+        sAuction.highestBid = auction.highestBid;
+        sAuction.highestBidder = auction.highestBidder;
+        if (firstBid) sAuction.startTime = auction.startTime;
+        if (durationExtended) sAuction.duration = auction.duration;
```

Comment:
is this cheaper than just writing the full auction struct to storage and having the fields on the auction in memory vs setting flags?

Reply:
I thought about this. I'm fine writing the entire struct but as the struct isn't really packed, I thought this would save gas. Would be quick to do an AB test